lib/view_model/active_record/update_operation.rb

```diff
@@ -526,35 +540,121 @@ def build_updates_for_has_many_through(association_data, reference_strings, upda
       previous_through_viewmodels_by_indirect_ref =
         previous_through_viewmodels.group_by(&viewmodel_reference_for_indirect_model)
 
-      new_through_viewmodels = reference_strings.map do |ref|
-        target_reference = update_context.resolve_reference(ref).viewmodel_reference
+      case association_update
+      when ReferencedCollectionUpdate::Replace
+        target_indirect_viewmodels =
+          association_update.values.map do |ref|
+            {
+              ref:           ref,
+              viewmodel_ref: update_context.resolve_reference(ref).viewmodel_reference,
+            }
+          end
+
+      when ReferencedCollectionUpdate::Functional
+        target_indirect_viewmodels =
+          previous_through_viewmodels.map { |vm| { viewmodel_ref: viewmodel_reference_for_indirect_model.(vm) } }
+
+        functional_update_check_duplicates!(association_update, update_context)
+
+        association_update.values.each do |fupdate|
+          case fupdate
+          when FunctionalUpdate::Append
+            appended_targets = fupdate.values.map do |ref|
+              {
+                ref:           ref,
+                viewmodel_ref: update_context.resolve_reference(ref).viewmodel_reference
+              }
+            end
+
+            if fupdate.before || fupdate.after
+              moved_refs =
+                fupdate.values.map { |ref| update_context.resolve_reference(ref).viewmodel_reference }.to_set
```

Comment:
I have an unrelated data structure that happens to contain the same values. Changed to digging them out instead of recomputing.